# Calculate error rates from Sep/17-Sep/21 in office testing. 
Testing build Survey2026 build #70 7192ca2. Built on 2026-09-16T16:13:20Z
https://github.com/ishizuki-tech/Survey2026/releases?page=1#release-build-70-7192ca2

In [1]:
import json

In [2]:
#define the path to the reference text
# One entry per survey recording: survey_id, question_id, expected_answer, wav_file, local_wav
reference_file = "office_testing/expected_answers.json"
reference = reference_file

# Read from a relative path
with open(reference, 'r', encoding='utf-8') as g:
    ref_data = json.load(g)

# Look up the expected answer by (survey_id, question_id), since a survey can have its own reference text
ref_lookup = {(r["survey_id"], r["question_id"]): r["expected_answer"] for r in ref_data}

print (f"Reference data: loaded {len(ref_data)} entries from {reference_file}")
for r in ref_data:
    print(f"{r['survey_id'][:8]} {r['question_id']}: {r['expected_answer']}")

Reference data: loaded 30 entries from office_testing/expected_answers.json
46ec4547 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
46ec4547 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
46ec4547 Q10: Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina.
46ec4547 Q11: Iwe na mavuno mengi na iwe inastahimili ukame na wadudu.
46ec4547 Q12: Hata tani moja kwa ekari, kama mbegu ni nzuri.
46ec4547 Q13: Wakati wa maua na kutengeneza punje.
46ec4547 Q14: Naweka gunia tano ya chakula, gunia tano nauza, na mbili ni ya mifugo.
46ec4547 Q15: Nawapa kuku na nguruwe mahindi mara mbili hadi tatu kwa wiki.
46ec4547 Q16: Ninanunua kwa agrovet kwa sababu ninaamini ubora wake.
46ec4547 Q17: Nauza kwa wafanyabiashara wa eneo kwa sababu wako karibu na hulipa haraka.
4f629181 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
4f629181 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
4f629181 Q10: Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina

In [3]:
# Define the GitHub URLs of the surveys we want to test today.
# Paste the normal github.com "blob" links; they are converted to raw file URLs below
survey_urls = [
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-17/exports/survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json",
]

from urllib.request import urlopen

# One entry per survey: file name plus the reference/answer pairs that matched.
surveys = []
for survey_url in survey_urls:
    raw_url = survey_url.replace("https://github.com/", "https://raw.githubusercontent.com/", 1).replace("/blob/", "/", 1)
    with urlopen(raw_url) as f:
        survey_data = json.load(f)

    survey_file = raw_url.rsplit("/", 1)[-1]
    print(f"\nLoaded data from {survey_file}")

    # Free text entered for the session (e.g. who tested and when), if present.
    session_free_text = survey_data.get('meta', {}).get('session_free_text', '')
    print(f"Session free text: {session_free_text}")

    # Keep only the questions that have a reference answer for this survey, in question order.
    survey_id = survey_data['survey_id']
    ref_questions = sorted((q for (sid, q) in ref_lookup if sid == survey_id), key=lambda q: int(q[1:]))
    if not ref_questions:
        print(f"No reference entries for survey {survey_id}, skipped")
        continue
    questions = [q for q in ref_questions if q in survey_data['answers']]
    missing = [q for q in ref_questions if q not in survey_data['answers']]
    if missing:
        print(f"Not in survey, skipped: {missing}")

    references = [ref_lookup[(survey_id, q)] for q in questions]
    answers = [survey_data['answers'][q]['answer'] for q in questions]
    print(f"Found {len(answers)} transcribed answers matching the reference.")
    for q, ans in zip(questions, answers):
        print(f"{q}: {ans}")

    surveys.append({"file": survey_file, "references": references, "answers": answers})


Loaded data from survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json
Session free text: This is Victor Kitoto filling in the survey on 17th September 2026
Found 10 transcribed answers matching the reference.
Q8: Ndiyo, elia ribu majani na kupunguza mawuno.
Q9: Na eza kukubali kupoteza hadia silimi ya kumi ya mawono.
Q10: Ikiva huwari defu inafika karibu asile miatela tini, intaba deleisha maa aina.
Q11: Iwe na mawuna mengi na iwe inasta emili ukame na wadulu
Q12: Atata ni moja kwaekari kama mbegunin zori.
Q13: Wakati wamaoa na kutengeneza punje
Q14: Ninaweka ya chakula na uza iliobaki na kidobu ni amifugo.
Q15: Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
 Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
Q16: Nina nunua kwa agrovet kwa sababu nina aminiu borawake.
Q17: Na uza kwa fanya bia shara wa ineu kwa sababu wa kukaribu na uli paraka.

Loaded data from survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json
Session free text:

In [4]:
import jiwer
import re
from jiwer.transforms import AbstractTransform

# Swahili number words, so "30", "30%" and "thelathini" all score the same.
_UNITS = ["", "moja", "mbili", "tatu", "nne", "tano", "sita", "saba", "nane", "tisa"]
_TENS = ["", "kumi", "ishirini", "thelathini", "arobaini", "hamsini", "sitini", "sabini", "themanini", "tisini"]

def swahili_number(n):
    """Spell out 0-999,999 in Swahili, e.g. 30 -> 'thelathini', 25 -> 'ishirini na tano'."""
    if n == 0:
        return "sifuri"
    parts = []
    if n >= 1000:
        parts.append("elfu " + swahili_number(n // 1000)); n %= 1000
    if n >= 100:
        parts.append("mia " + swahili_number(n // 100)); n %= 100
    if n >= 10:
        parts.append(_TENS[n // 10]); n %= 10
    if n:
        parts.append(_UNITS[n])
    return " na ".join(parts)

class NormalizeSwahiliNumbers(AbstractTransform):
    """Digits to Swahili words: '2-3' -> 'mbili hadi tatu', '10%' -> 'asilimia kumi'."""
    def process_string(self, s):
        num = lambda m: swahili_number(int(m))
        s = re.sub(r"\b(\d+)\s*[-–]\s*(\d+)\b", lambda m: f"{num(m[1])} hadi {num(m[2])}", s)  # ranges
        s = re.sub(r"\b(?:asilimia\s+)?(\d+)\s*%", lambda m: f"asilimia {num(m[1])}", s)       # percentages
        return re.sub(r"\b\d+\b", lambda m: num(m[0]), s)                                      # plain numbers


transform_words = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfWords()
])

transform_chars = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfChars()
])

def error_rates(references, answers):
    out = jiwer.process_words(references, answers, reference_transform=transform_words, hypothesis_transform=transform_words)
    out2 = jiwer.process_characters(references, answers, reference_transform=transform_chars, hypothesis_transform=transform_chars)
    return out.wer, out2.cer

# Per-survey rates.
for s in surveys:
    wer, cer = error_rates(s["references"], s["answers"])
    print(f"{s['file']}  ({len(s['answers'])} answers)")
    print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

# Aggregate over every answer from every survey: total errors / total reference words (or characters).
all_references = [r for s in surveys for r in s["references"]]
all_answers = [a for s in surveys for a in s["answers"]]
wer, cer = error_rates(all_references, all_answers)
print(f"\nAggregate over {len(surveys)} surveys, {len(all_answers)} answers")
print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json  (10 answers)
  WER (normalized): 0.780   CER (normalized): 0.296
survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json  (10 answers)
  WER (normalized): 0.659   CER (normalized): 0.120
survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json  (10 answers)
  WER (normalized): 0.727   CER (normalized): 0.120

Aggregate over 3 surveys, 30 answers
  WER (normalized): 0.723   CER (normalized): 0.179


## Reference text and extra speech

Surveys from Sep/17–Sep/21: 30 scored answers (Q8–Q17 from each of the 3 surveys). Reference text comes from `office_testing/expected_answers.json`. It has one entry per survey and question, and each entry links to its recording.

- Numbers are spelled out ("asilimia kumi", "tani moja"), matching how they are spoken.
- Where a question was recorded more than once, the entry points to the last take, by filename timestamp.

### Corrections to the scripted answers

Two references for the 2026-09-17 survey (`46ec4547`, Victor Kitoto) were changed to match what is actually in the recording. The basis is transcriptions by five independent models (four Swahili Whisper models and omniASR-LLM-7B).

| Question | Reference now | Why |
|---|---|---|
| Q14 | "Naweka gunia tano ya chakula, gunia tano nauza, na mbili ni ya mifugo." | The last of three takes is a different answer from the script. The first two takes are the scripted answer. All five models agree on this wording. |
| Q15 | The scripted answer, once | The app transcript shows the answer twice, but no model hears a second clear repeat. The 39.3 s recording has the answer at 0–6 s, near-silence, then quieter, unclear sound at 0:30–0:39. |

### Effect on build #70

| | WER | CER |
|---|---|---|
| 2026-09-17, corrected references | 0.780 | 0.296 |
| **Aggregate, corrected references** | **0.723** | **0.179** |

The 2026-09-17 survey scores worse than before, for two reasons:

- **Q15:** the app's doubled Q15 transcript now counts as inserted words.
- **Q14:** the app's saved Q14 transcript comes from an earlier take, which is the scripted answer, not the last take the reference follows. Scored against the take it actually transcribed, the aggregate is WER 0.720 / CER 0.163.

The two Sep/21 surveys are unchanged: WER 0.659 / CER 0.120 for `4f629181`, and WER 0.727 / CER 0.120 for `7db59ab6`.

No other extra speech shows up in the transcripts. One recording is long for its question: `4f629181` Q15 at 14.2 s against about 8 s elsewhere. All models transcribe it as the plain answer.

## Swahili model comparison (Whisper and omniASR)

The same 30 recordings, transcribed by the four Swahili fine-tuned Whisper Small models from the Survey2026 `feature/swahili-whisper-*` branches, and by five omniASR models. Transcripts are in `office_testing/model_transcripts.json`.

- Runtime: whisper.cpp v1.9.3 (`371b5a7`, the version pinned in Survey2026), on a cloud CPU rather than a phone.
- Decoding matches the app's `whisperLib.c`: language `sw`, greedy (best_of 1), no context, no timestamps.
- Models were taken from the branches' Git LFS files and checked against the SHA-256 values in `SWAHILI_WHISPER_MODEL_COMPARISON.md`. The preview APKs can't be used for this, because they contain LFS pointer stubs instead of the models.

Five Meta Omnilingual ASR models were run separately on a Colab A100 with `omniASR_office_testing.ipynb`: bf16, language `swh_Latn`, default decoding. Their transcripts are in `office_testing/omniasr_transcripts.json`.

| Model | Parameters | Download | GPU memory |
|---|---|---|---|
| `omniASR_LLM_7B` | ~7.8B | 31 GB | ~17 GiB |
| `omniASR_LLM_300M`, `_v2` | 1.6B (300M encoder + LLM decoder) | 6.1 GiB | ~5 GiB |
| `omniASR_CTC_300M`, `_v2` | 325M | 1.3 GiB | ~2 GiB |

Of these, only the CTC models are close to Whisper Small's size (244M parameters, 175 MB quantized). omniASR-LLM-7B writes some numbers as digits ("asilimia 30", "mara 2-3"). Scoring converts digits, percentages and ranges to Swahili words in both the reference and every transcript (`NormalizeSwahiliNumbers` in the scoring cell), so number format isn't counted as an error. A 7B model is far too large to run on a phone.

Build #70 is the app's on-phone result with the generic `ggml-small-q5_1` model, so that row isn't run under identical conditions.

Build #70's Q14 for `46ec4547` is scored against a different take than the models (see above), which costs it about 0.016 CER.

In [5]:
# Score each model with the same transforms and reference text as above
# Whisper models (whisper.cpp) and omniASR (Colab) transcripts of the same 30 recordings
model_data = {}
for path in ["office_testing/model_transcripts.json", "office_testing/omniasr_transcripts.json"]:
    with open(path, encoding="utf-8") as f:
        model_data.update(json.load(f)["models"])

survey_ids = sorted({sid for (sid, q) in ref_lookup})

def score(transcripts):
    """transcripts: list of dicts with survey_id, question_id, transcript"""
    by_key = {(t["survey_id"], t["question_id"]): t["transcript"] for t in transcripts}
    keys = sorted(by_key, key=lambda k: (survey_ids.index(k[0]), int(k[1][1:])))
    total = error_rates([ref_lookup[k] for k in keys], [by_key[k] for k in keys])
    per_survey = [error_rates([ref_lookup[k] for k in keys if k[0] == sid],
                              [by_key[k] for k in keys if k[0] == sid]) for sid in survey_ids]
    return len(keys), total, per_survey

rows = [("Build #70 app (generic small)", len(all_answers), error_rates(all_references, all_answers),
         [error_rates(s["references"], s["answers"]) for s in surveys])]
for name, m in model_data.items():
    rows.append((name, *score(m["transcripts"])))

header = f"{'Model':32s} {'n':>3s}  {'WER':>6s} {'CER':>6s}   " + "   ".join(f"{sid[:8]} WER/CER" for sid in survey_ids)
print(header)
print("-" * len(header))
for name, n, (wer, cer), per in rows:
    print(f"{name:32s} {n:3d}  {wer:6.3f} {cer:6.3f}   " + "   ".join(f"{w:6.3f}/{c:.3f}  " for w, c in per))

Model                              n     WER    CER   46ec4547 WER/CER   4f629181 WER/CER   7db59ab6 WER/CER
------------------------------------------------------------------------------------------------------------
Build #70 app (generic small)     30   0.723  0.179    0.780/0.296      0.659/0.120      0.727/0.120  
lingua-connect                    30   0.446  0.094    0.516/0.103      0.432/0.094      0.386/0.083  
cdli-cv20                         30   0.273  0.054    0.319/0.074      0.227/0.048      0.273/0.041  
cdli-kenyan-nonstandard           30   0.236  0.062    0.286/0.087      0.239/0.050      0.182/0.048  
dmusingu-cv14                     30   0.146  0.061    0.231/0.142      0.091/0.015      0.114/0.024  
omniASR_LLM_7B                    30   0.071  0.021    0.055/0.011      0.034/0.018      0.125/0.033  
omniASR_LLM_300M                  30   0.116  0.038    0.165/0.047      0.057/0.009      0.125/0.057  
omniASR_LLM_300M_v2               30   0.165  0.026    0.176/